File: backend/app/services/forms_assistant.py
from typing import Any


def normalize_language(language: str | None) -> str:
    lang = (language or "en").strip().lower()
    return "fr" if lang == "fr" else "en"


def t(en: str, fr: str, language: str) -> str:
    return fr if language == "fr" else en


def yes_no(value: bool, language: str) -> str:
    return t("Yes", "Oui", language) if value else t("No", "Non", language)


def build_study_permit_forms_assistant(
    intake: dict[str, Any] | None = None,
    language: str = "en",
) -> dict[str, Any]:
    language = normalize_language(language)
    intake = intake or {}

    missing_fields: list[str] = []
    recommended_forms: list[dict[str, Any]] = [
        {
            "form_key": "study_permit_main",
            "form_name": t(
                "Study permit application package",
                "Dossier de demande de permis d’études",
                language,
            ),
            "status": "Primary",
            "notes": t(
                "Primary study permit workflow for the applicant.",
                "Processus principal de demande de permis d’études pour le demandeur.",
                language,
            ),
        },
        {
            "form_key": "family_information",
            "form_name": t(
                "Family information form",
                "Formulaire d’information familiale",
                language,
            ),
            "status": "Supporting",
            "notes": t(
                "Often required where family and background details must be disclosed.",
                "Souvent requis lorsque les renseignements familiaux et biographiques doivent être déclarés.",
                language,
            ),
        },
    ]

    draft_answers = {
        "school_or_dli": intake.get("dli_name") or intake.get("school_name") or "",
        "program_name": intake.get("program_name") or "",
        "intake_term": intake.get("intake_term") or "",
        "tuition_amount": intake.get("tuition_amount") or "",
        "proof_of_funds_available": yes_no(
            intake.get("proof_of_funds_available"), language
        ),
        "previous_refusal": yes_no(intake.get("previous_refusal"), language),
        "accompanying_family": yes_no(intake.get("accompanying_family"), language),
        "passport_valid": yes_no(intake.get("passport_valid"), language),
        "study_gap_explanation": intake.get("gap_in_studies_explanation") or "",
    }

    if not draft_answers["school_or_dli"]:
        missing_fields.append(
            t(
                "School or DLI name",
                "Nom de l’établissement ou de l’EED",
                language,
            )
        )
    if not draft_answers["program_name"]:
        missing_fields.append(t("Program name", "Nom du programme", language))
    if not draft_answers["intake_term"]:
        missing_fields.append(t("Intake term", "Session d’entrée", language))
    if not draft_answers["tuition_amount"]:
        missing_fields.append(
            t("Tuition amount", "Montant des frais de scolarité", language)
        )
    if draft_answers["proof_of_funds_available"] == yes_no(False, language):
        missing_fields.append(
            t("Proof of funds strategy", "Stratégie de preuve de fonds", language)
        )
    if draft_answers["passport_valid"] == yes_no(False, language):
        missing_fields.append(
            t(
                "Passport validity confirmation",
                "Confirmation de la validité du passeport",
                language,
            )
        )

    preparation_notes = [
        t(
            "Review identity, travel, and education details together so the file reads consistently from start to finish.",
            "Revoyez ensemble les renseignements d’identité, de voyage et d’études afin que le dossier soit cohérent du début à la fin.",
            language,
        ),
        t(
            "Make sure the study plan clearly explains school choice, program relevance, and realistic future plans.",
            "Assurez-vous que le projet d’études explique clairement le choix de l’établissement, la pertinence du programme et des objectifs futurs réalistes.",
            language,
        ),
        t(
            "Align proof of funds with tuition amounts and expected living costs so the financial section feels credible.",
            "Alignez la preuve de fonds avec les frais de scolarité et le coût de la vie prévu afin que la section financière soit crédible.",
            language,
        ),
    ]

    if intake.get("previous_refusal"):
        preparation_notes.append(
            t(
                "Prepare a focused refusal-response explanation that addresses earlier concerns directly and calmly.",
                "Préparez une explication ciblée au refus antérieur qui répond directement et clairement aux préoccupations précédentes.",
                language,
            )
        )

    if intake.get("accompanying_family"):
        recommended_forms.append(
            {
                "form_key": "family_member_support",
                "form_name": t(
                    "Accompanying family support package",
                    "Dossier complémentaire pour la famille accompagnante",
                    language,
                ),
                "status": "Conditional",
                "notes": t(
                    "Additional family-related forms and supporting information may be required.",
                    "Des formulaires supplémentaires et des pièces justificatives liées à la famille peuvent être requis.",
                    language,
                ),
            }
        )

    summary = (
        t(
            "The study permit forms package looks well-positioned from the current intake. The next priority should be polish, consistency, and strong supporting evidence.",
            "Le dossier de formulaires de permis d’études paraît bien positionné selon les informations actuelles. La prochaine priorité devrait être la qualité de présentation, la cohérence et la solidité des pièces justificatives.",
            language,
        )
        if len(missing_fields) == 0
        else t(
            "The forms package has a workable base, but several important items still need to be completed or reviewed before it feels submission-ready.",
            "Le dossier de formulaires repose sur une base utilisable, mais plusieurs éléments importants doivent encore être complétés ou vérifiés avant qu’il paraisse prêt à être soumis.",
            language,
        )
    )

    return {
        "matter_type": "study_permit",
        "package_title": t(
            "Study Permit Forms Assistant",
            "Assistant de formulaires – Permis d’études",
            language,
        ),
        "recommended_forms": recommended_forms,
        "draft_answers": draft_answers,
        "missing_fields": missing_fields,
        "preparation_notes": preparation_notes,
        "summary": summary,
    }


def build_work_permit_forms_assistant(
    intake: dict[str, Any] | None = None,
    language: str = "en",
) -> dict[str, Any]:
    language = normalize_language(language)
    intake = intake or {}

    missing_fields: list[str] = []
    recommended_forms: list[dict[str, Any]] = [
        {
            "form_key": "work_permit_main",
            "form_name": t(
                "Work permit application package",
                "Dossier de demande de permis de travail",
                language,
            ),
            "status": "Primary",
            "notes": t(
                "Primary work permit workflow for the applicant.",
                "Processus principal de demande de permis de travail pour le demandeur.",
                language,
            ),
        },
        {
            "form_key": "family_information",
            "form_name": t(
                "Family information form",
                "Formulaire d’information familiale",
                language,
            ),
            "status": "Supporting",
            "notes": t(
                "Often required where family and background details must be disclosed.",
                "Souvent requis lorsque les renseignements familiaux et biographiques doivent être déclarés.",
                language,
            ),
        },
    ]

    permit_type = intake.get("permit_type") or ""

    draft_answers = {
        "permit_type": permit_type,
        "employer_name": intake.get("employer_name") or "",
        "job_title": intake.get("job_title") or "",
        "noc_code": intake.get("noc_code") or "",
        "province_of_work": intake.get("province_of_work") or "",
        "wage": intake.get("wage") or "",
        "current_status_in_canada": intake.get("current_status_in_canada") or "",
        "expires_on": intake.get("expires_on") or "",
        "lmia_available": yes_no(intake.get("lmia_available"), language),
        "open_work_permit_basis": intake.get("open_work_permit_basis") or "",
        "accompanying_family": yes_no(intake.get("accompanying_family"), language),
    }

    if not draft_answers["permit_type"]:
        missing_fields.append(t("Permit type", "Type de permis", language))
    if not draft_answers["job_title"]:
        missing_fields.append(t("Job title", "Titre du poste", language))
    if not draft_answers["current_status_in_canada"]:
        missing_fields.append(
            t("Current status in Canada", "Statut actuel au Canada", language)
        )

    lowered_permit_type = permit_type.lower()

    if "employer" in lowered_permit_type and not draft_answers["employer_name"]:
        missing_fields.append(t("Employer name", "Nom de l’employeur", language))

    if "employer" in lowered_permit_type and draft_answers["lmia_available"] == yes_no(False, language):
        missing_fields.append(
            t("LMIA or exemption confirmation", "Confirmation EIMT ou exemption", language)
        )

    if "open" in lowered_permit_type and not draft_answers["open_work_permit_basis"]:
        missing_fields.append(
            t("Open work permit basis", "Fondement du permis ouvert", language)
        )

    preparation_notes = [
        t(
            "Confirm that the employer, job title, and NOC details remain consistent across all supporting documents.",
            "Confirmez que l’employeur, le titre du poste et les détails du CNP demeurent cohérents dans l’ensemble des pièces justificatives.",
            language,
        ),
        t(
            "Review the applicant’s current immigration status and expiry timeline carefully before finalizing the filing sequence.",
            "Examinez attentivement le statut actuel du demandeur et l’échéancier d’expiration avant de finaliser l’ordre de dépôt.",
            language,
        ),
        t(
            "Confirm early whether the case is LMIA-based or LMIA-exempt so the file strategy is built on the correct procedural basis.",
            "Confirmez tôt si le dossier repose sur une EIMT ou sur une exemption afin de bâtir la stratégie sur le bon fondement procédural.",
            language,
        ),
    ]

    if intake.get("accompanying_family"):
        recommended_forms.append(
          {
              "form_key": "family_member_support",
              "form_name": t(
                  "Accompanying family support package",
                  "Dossier complémentaire pour la famille accompagnante",
                  language,
              ),
              "status": "Conditional",
              "notes": t(
                  "Additional family-related forms and supporting information may be required.",
                  "Des formulaires supplémentaires et des pièces justificatives liées à la famille peuvent être requis.",
                  language,
              ),
          }
        )

    summary = (
        t(
            "The work permit forms package looks well-positioned from the current intake. The next priority should be procedural accuracy and consistency across supporting records.",
            "Le dossier de formulaires de permis de travail paraît bien positionné selon les informations actuelles. La prochaine priorité devrait être la précision procédurale et la cohérence des pièces justificatives.",
            language,
        )
        if len(missing_fields) == 0
        else t(
            "The work permit forms package has a workable base, but several important items still need to be completed or clarified before the file feels strong.",
            "Le dossier de formulaires de permis de travail repose sur une base utilisable, mais plusieurs éléments importants doivent encore être complétés ou clarifiés avant que le dossier paraisse solide.",
            language,
        )
    )

    return {
        "matter_type": "work_permit",
        "package_title": t(
            "Work Permit Forms Assistant",
            "Assistant de formulaires – Permis de travail",
            language,
        ),
        "recommended_forms": recommended_forms,
        "draft_answers": draft_answers,
        "missing_fields": missing_fields,
        "preparation_notes": preparation_notes,
        "summary": summary,
    }


def build_spousal_sponsorship_forms_assistant(
    intake: dict[str, Any] | None = None,
    language: str = "en",
) -> dict[str, Any]:
    language = normalize_language(language)
    intake = intake or {}

    missing_fields: list[str] = []
    recommended_forms: list[dict[str, Any]] = [
        {
            "form_key": "sponsorship_package_main",
            "form_name": t(
                "Spousal sponsorship application package",
                "Dossier principal de parrainage conjugal",
                language,
            ),
            "status": "Primary",
            "notes": t(
                "Primary sponsorship workflow for the sponsor and principal applicant.",
                "Processus principal de parrainage pour le répondant et le demandeur principal.",
                language,
            ),
        },
        {
            "form_key": "relationship_history",
            "form_name": t(
                "Relationship history and evidence package",
                "Dossier d’historique et de preuves de la relation",
                language,
            ),
            "status": "Supporting",
            "notes": t(
                "Used to organize relationship timeline and supporting evidence.",
                "Utilisé pour organiser la chronologie relationnelle et les pièces justificatives.",
                language,
            ),
        },
    ]

    relationship_type = intake.get("relationship_type") or ""

    draft_answers = {
        "sponsor_status": intake.get("sponsor_status") or "",
        "relationship_type": relationship_type,
        "relationship_start_date": intake.get("relationship_start_date") or "",
        "marriage_date": intake.get("marriage_date") or "",
        "cohabiting": yes_no(intake.get("cohabiting"), language),
        "principal_applicant_country": intake.get("principal_applicant_country") or "",
        "dependent_children": yes_no(intake.get("dependent_children"), language),
        "previous_marriage_or_sponsorship": yes_no(
            intake.get("previous_marriage_or_sponsorship"), language
        ),
        "police_certificates_ready": yes_no(
            intake.get("police_certificates_ready"), language
        ),
        "medicals_ready": yes_no(intake.get("medicals_ready"), language),
        "proof_of_relationship_notes": intake.get("proof_of_relationship_notes") or "",
    }

    if not draft_answers["sponsor_status"]:
        missing_fields.append(t("Sponsor status", "Statut du répondant", language))
    if not draft_answers["relationship_type"]:
        missing_fields.append(t("Relationship type", "Type de relation", language))
    if not draft_answers["relationship_start_date"]:
        missing_fields.append(
            t("Relationship start date", "Date de début de la relation", language)
        )
    if "spouse" in relationship_type.lower() and not draft_answers["marriage_date"]:
        missing_fields.append(t("Marriage date", "Date du mariage", language))
    if not draft_answers["proof_of_relationship_notes"]:
        missing_fields.append(
            t("Relationship evidence notes", "Notes sur les preuves relationnelles", language)
        )
    if draft_answers["police_certificates_ready"] == yes_no(False, language):
        missing_fields.append(
            t("Police certificate planning", "Planification des certificats de police", language)
        )
    if draft_answers["medicals_ready"] == yes_no(False, language):
        missing_fields.append(
            t("Medical exam planning", "Planification de l’examen médical", language)
        )

    preparation_notes = [
        t(
            "Ensure the relationship history remains consistent across all forms, declarations, and supporting evidence.",
            "Assurez-vous que l’historique de la relation demeure cohérent dans tous les formulaires, déclarations et éléments de preuve.",
            language,
        ),
        t(
            "Prepare a clear relationship timeline covering communication, visits, cohabitation, and major milestones.",
            "Préparez une chronologie claire de la relation couvrant les communications, les visites, la cohabitation et les étapes importantes.",
            language,
        ),
        t(
            "Review any prior marriage, sponsorship, or dependent child issues carefully so they are documented in a stable and credible way.",
            "Examinez attentivement tout mariage antérieur, historique de parrainage ou question liée aux enfants à charge afin qu’ils soient documentés de façon stable et crédible.",
            language,
        ),
    ]

    if intake.get("dependent_children"):
        recommended_forms.append(
            {
                "form_key": "dependent_children_support",
                "form_name": t(
                    "Dependent children support package",
                    "Dossier complémentaire pour enfants à charge",
                    language,
                ),
                "status": "Conditional",
                "notes": t(
                    "Additional child-related supporting information may be required.",
                    "Des pièces justificatives supplémentaires liées aux enfants peuvent être requises.",
                    language,
                ),
            }
        )

    summary = (
        t(
            "The sponsorship forms package looks well-positioned from the current intake. The next priority should be evidence quality, timeline clarity, and overall credibility.",
            "Le dossier de formulaires de parrainage paraît bien positionné selon les informations actuelles. La prochaine priorité devrait être la qualité des preuves, la clarté chronologique et la crédibilité globale.",
            language,
        )
        if len(missing_fields) == 0
        else t(
            "The sponsorship forms package has a workable base, but several important relationship and procedural details still need to be completed or reinforced.",
            "Le dossier de formulaires de parrainage repose sur une base utilisable, mais plusieurs éléments relationnels et procéduraux importants doivent encore être complétés ou renforcés.",
            language,
        )
    )

    return {
        "matter_type": "spousal_sponsorship",
        "package_title": t(
            "Spousal Sponsorship Forms Assistant",
            "Assistant de formulaires – Parrainage du conjoint",
            language,
        ),
        "recommended_forms": recommended_forms,
        "draft_answers": draft_answers,
        "missing_fields": missing_fields,
        "preparation_notes": preparation_notes,
        "summary": summary,
    }


def build_forms_assistant(
    matter_type: str | None,
    intake: dict[str, Any] | None = None,
    language: str = "en",
) -> dict[str, Any]:
    language = normalize_language(language)

    if matter_type == "study_permit":
        return build_study_permit_forms_assistant(intake, language)

    if matter_type == "work_permit":
        return build_work_permit_forms_assistant(intake, language)

    if matter_type == "spousal_sponsorship":
        return build_spousal_sponsorship_forms_assistant(intake, language)

    return {
        "matter_type": matter_type or "unknown",
        "package_title": t("Forms Assistant", "Assistant de formulaires", language),
        "recommended_forms": [],
        "draft_answers": {},
        "missing_fields": [
            t(
                "Forms assistant has not been built for this matter type yet.",
                "L’assistant de formulaires n’a pas encore été créé pour ce type de dossier.",
                language,
            )
        ],
        "preparation_notes": [],
        "summary": t(
            "No forms assistant is available for this matter type yet.",
            "Aucun assistant de formulaires n’est encore disponible pour ce type de dossier.",
            language,
        ),
    }

File: frontend/src/utils/formsAssistant.js
function normalizeFormsLanguage(language) {
  return String(language || "en").toLowerCase().startsWith("fr") ? "fr" : "en";
}

const formsAssistantFr = {
  Yes: "Oui",
  No: "Non",
  Primary: "Principal",
  Supporting: "Appui",
  Conditional: "Conditionnel",
  "Study Permit Forms Assistant": "Assistant formulaires - permis d'etudes",
  "Work Permit Forms Assistant": "Assistant formulaires - permis de travail",
  "Spousal Sponsorship Forms Assistant": "Assistant formulaires - parrainage d'epoux/conjoint",
  "Forms Assistant": "Assistant formulaires",
  "Study permit application package": "Dossier de demande de permis d'etudes",
  "Family information form": "Formulaire de renseignements sur la famille",
  "Accompanying family support package": "Dossier d'appui pour la famille accompagnante",
  "Work permit application package": "Dossier de demande de permis de travail",
  "Spousal sponsorship application package": "Dossier de parrainage d'epoux/conjoint",
  "Relationship history and evidence package": "Dossier d'historique de relation et de preuves",
  "Dependent children support package": "Dossier d'appui pour enfants a charge",
  "Main study permit application workflow for the applicant.":
    "Flux principal de demande de permis d'etudes pour le demandeur.",
  "Usually needed where family/background details are required.":
    "Habituellement requis lorsque des renseignements familiaux ou personnels sont demandes.",
  "Family-related forms and supporting information may be required.":
    "Des formulaires et renseignements d'appui lies a la famille peuvent etre requis.",
  "Main work permit workflow for the applicant.":
    "Flux principal de demande de permis de travail pour le demandeur.",
  "Main sponsorship workflow for sponsor and principal applicant.":
    "Flux principal de parrainage pour le repondant et le demandeur principal.",
  "Relationship timeline and supporting evidence planning.":
    "Planification de la chronologie de relation et des preuves d'appui.",
  "Additional child-related supporting information may be required.":
    "Des renseignements d'appui supplementaires lies aux enfants peuvent etre requis.",
  "School or DLI name": "Nom de l'ecole ou de l'EED",
  "Program name": "Nom du programme",
  "Intake term": "Session d'admission",
  "Tuition amount": "Montant des frais de scolarite",
  "Proof of funds strategy": "Strategie de preuve de fonds",
  "Passport validity confirmation": "Confirmation de validite du passeport",
  "Permit type": "Type de permis",
  "Job title": "Titre du poste",
  "Current status in Canada": "Statut actuel au Canada",
  "Employer name": "Nom de l'employeur",
  "LMIA or exemption confirmation": "Confirmation d'EIMT ou d'exemption",
  "Open work permit basis": "Base du permis de travail ouvert",
  "Sponsor status": "Statut du repondant",
  "Relationship type": "Type de relation",
  "Relationship start date": "Date de debut de la relation",
  "Marriage date": "Date du mariage",
  "Relationship evidence notes": "Notes sur les preuves de relation",
  "Police certificate planning": "Planification des certificats de police",
  "Medical exam planning": "Planification de l'examen medical",
  "Review all identity, travel, and education details for consistency.":
    "Verifier la coherence de l'identite, des voyages et des etudes.",
  "Make sure the study plan clearly explains school choice, program relevance, and future plans.":
    "S'assurer que le plan d'etudes explique clairement le choix de l'ecole, la pertinence du programme et les projets futurs.",
  "Match proof of funds with tuition and living cost expectations.":
    "Faire correspondre la preuve de fonds aux frais de scolarite et aux couts de subsistance prevus.",
  "Prepare a clear refusal-response explanation addressing prior concerns.":
    "Preparer une reponse claire aux refus precedents et aux preoccupations soulevees.",
  "Confirm job, employer, and NOC details are consistent across documents.":
    "Confirmer la coherence de l'emploi, de l'employeur et du NOC dans tous les documents.",
  "Review the applicant's current immigration status and expiry timeline carefully.":
    "Verifier attentivement le statut d'immigration actuel et les dates d'expiration.",
  "Confirm whether this case is LMIA-based or LMIA-exempt where relevant.":
    "Confirmer si le dossier repose sur une EIMT ou sur une exemption, le cas echeant.",
  "Ensure relationship history is consistent across all forms and evidence.":
    "S'assurer que l'historique de la relation est coherent dans les formulaires et les preuves.",
  "Prepare a clean timeline of communication, visits, cohabitation, and major milestones.":
    "Preparer une chronologie claire des communications, visites, cohabitation et evenements importants.",
  "Review prior marriages, sponsorships, or dependent child issues carefully.":
    "Verifier attentivement les mariages anterieurs, parrainages precedents et questions liees aux enfants a charge.",
  "The study permit forms package looks reasonably prepared from current intake.":
    "Le dossier de permis d'etudes semble raisonnablement prepare selon les renseignements actuels.",
  "Some key study permit form fields still need to be completed or reviewed.":
    "Certains champs importants du permis d'etudes doivent encore etre completes ou verifies.",
  "The work permit forms package looks reasonably prepared from current intake.":
    "Le dossier de permis de travail semble raisonnablement prepare selon les renseignements actuels.",
  "Some key work permit form fields still need to be completed or reviewed.":
    "Certains champs importants du permis de travail doivent encore etre completes ou verifies.",
  "The sponsorship forms package looks reasonably prepared from current intake.":
    "Le dossier de parrainage semble raisonnablement prepare selon les renseignements actuels.",
  "Some key sponsorship form fields still need to be completed or reviewed.":
    "Certains champs importants du parrainage doivent encore etre completes ou verifies.",
  "Forms assistant has not been built for this matter type yet.":
    "L'assistant formulaires n'est pas encore disponible pour ce type de dossier.",
  "No forms assistant is available for this matter type yet.":
    "Aucun assistant formulaires n'est encore disponible pour ce type de dossier.",
};

function translateFormsAssistantValue(value, language) {
  if (normalizeFormsLanguage(language) !== "fr") return value;

  if (typeof value === "string") {
    return formsAssistantFr[value] || value;
  }

  if (Array.isArray(value)) {
    return value.map((item) => translateFormsAssistantValue(item, language));
  }

  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [
        key,
        translateFormsAssistantValue(item, language),
      ])
    );
  }

  return value;
}

function buildStudyPermitFormsAssistant(intake = {}) {
  const missing_fields = [];
  const recommended_forms = [
    {
      form_key: "study_permit_main",
      form_name: "Study permit application package",
      status: "Primary",
      notes: "Main study permit application workflow for the applicant.",
    },
    {
      form_key: "family_information",
      form_name: "Family information form",
      status: "Supporting",
      notes: "Usually needed where family/background details are required.",
    },
  ];

  const draft_answers = {
    school_or_dli: intake.dli_name || intake.school_name || "",
    program_name: intake.program_name || "",
    intake_term: intake.intake_term || "",
    tuition_amount: intake.tuition_amount || "",
    proof_of_funds_available: intake.proof_of_funds_available ? "Yes" : "No",
    previous_refusal: intake.previous_refusal ? "Yes" : "No",
    accompanying_family: intake.accompanying_family ? "Yes" : "No",
    passport_valid: intake.passport_valid ? "Yes" : "No",
    study_gap_explanation: intake.gap_in_studies_explanation || "",
  };

  if (!draft_answers.school_or_dli) missing_fields.push("School or DLI name");
  if (!draft_answers.program_name) missing_fields.push("Program name");
  if (!draft_answers.intake_term) missing_fields.push("Intake term");
  if (!draft_answers.tuition_amount) missing_fields.push("Tuition amount");
  if (draft_answers.proof_of_funds_available === "No") {
    missing_fields.push("Proof of funds strategy");
  }
  if (draft_answers.passport_valid === "No") {
    missing_fields.push("Passport validity confirmation");
  }

  const preparation_notes = [
    "Review all identity, travel, and education details for consistency.",
    "Make sure the study plan clearly explains school choice, program relevance, and future plans.",
    "Match proof of funds with tuition and living cost expectations.",
  ];

  if (intake.previous_refusal) {
    preparation_notes.push(
      "Prepare a clear refusal-response explanation addressing prior concerns."
    );
  }

  if (intake.accompanying_family) {
    recommended_forms.push({
      form_key: "family_member_support",
      form_name: "Accompanying family support package",
      status: "Conditional",
      notes: "Family-related forms and supporting information may be required.",
    });
  }

  return {
    matter_type: "study_permit",
    package_title: "Study Permit Forms Assistant",
    recommended_forms,
    draft_answers,
    missing_fields,
    preparation_notes,
    summary:
      missing_fields.length === 0
        ? "The study permit forms package looks reasonably prepared from current intake."
        : "Some key study permit form fields still need to be completed or reviewed.",
  };
}

function buildWorkPermitFormsAssistant(intake = {}) {
  const missing_fields = [];
  const recommended_forms = [
    {
      form_key: "work_permit_main",
      form_name: "Work permit application package",
      status: "Primary",
      notes: "Main work permit workflow for the applicant.",
    },
    {
      form_key: "family_information",
      form_name: "Family information form",
      status: "Supporting",
      notes: "Usually needed where family/background details are required.",
    },
  ];

  const draft_answers = {
    permit_type: intake.permit_type || "",
    employer_name: intake.employer_name || "",
    job_title: intake.job_title || "",
    noc_code: intake.noc_code || "",
    province_of_work: intake.province_of_work || "",
    wage: intake.wage || "",
    current_status_in_canada: intake.current_status_in_canada || "",
    expires_on: intake.expires_on || "",
    lmia_available: intake.lmia_available ? "Yes" : "No",
    open_work_permit_basis: intake.open_work_permit_basis || "",
    accompanying_family: intake.accompanying_family ? "Yes" : "No",
  };

  if (!draft_answers.permit_type) missing_fields.push("Permit type");
  if (!draft_answers.job_title) missing_fields.push("Job title");
  if (!draft_answers.current_status_in_canada) {
    missing_fields.push("Current status in Canada");
  }

  if (
    draft_answers.permit_type.toLowerCase().includes("employer") &&
    !draft_answers.employer_name
  ) {
    missing_fields.push("Employer name");
  }

  if (
    draft_answers.permit_type.toLowerCase().includes("employer") &&
    draft_answers.lmia_available === "No"
  ) {
    missing_fields.push("LMIA or exemption confirmation");
  }

  if (
    draft_answers.permit_type.toLowerCase().includes("open") &&
    !draft_answers.open_work_permit_basis
  ) {
    missing_fields.push("Open work permit basis");
  }

  const preparation_notes = [
    "Confirm job, employer, and NOC details are consistent across documents.",
    "Review the applicant's current immigration status and expiry timeline carefully.",
    "Confirm whether this case is LMIA-based or LMIA-exempt where relevant.",
  ];

  if (intake.accompanying_family) {
    recommended_forms.push({
      form_key: "family_member_support",
      form_name: "Accompanying family support package",
      status: "Conditional",
      notes: "Family-related forms and supporting information may be required.",
    });
  }

  return {
    matter_type: "work_permit",
    package_title: "Work Permit Forms Assistant",
    recommended_forms,
    draft_answers,
    missing_fields,
    preparation_notes,
    summary:
      missing_fields.length === 0
        ? "The work permit forms package looks reasonably prepared from current intake."
        : "Some key work permit form fields still need to be completed or reviewed.",
  };
}

function buildSpousalSponsorshipFormsAssistant(intake = {}) {
  const missing_fields = [];
  const recommended_forms = [
    {
      form_key: "sponsorship_package_main",
      form_name: "Spousal sponsorship application package",
      status: "Primary",
      notes: "Main sponsorship workflow for sponsor and principal applicant.",
    },
    {
      form_key: "relationship_history",
      form_name: "Relationship history and evidence package",
      status: "Supporting",
      notes: "Relationship timeline and supporting evidence planning.",
    },
  ];

  const draft_answers = {
    sponsor_status: intake.sponsor_status || "",
    relationship_type: intake.relationship_type || "",
    relationship_start_date: intake.relationship_start_date || "",
    marriage_date: intake.marriage_date || "",
    cohabiting: intake.cohabiting ? "Yes" : "No",
    principal_applicant_country: intake.principal_applicant_country || "",
    dependent_children: intake.dependent_children ? "Yes" : "No",
    previous_marriage_or_sponsorship: intake.previous_marriage_or_sponsorship
      ? "Yes"
      : "No",
    police_certificates_ready: intake.police_certificates_ready ? "Yes" : "No",
    medicals_ready: intake.medicals_ready ? "Yes" : "No",
    proof_of_relationship_notes: intake.proof_of_relationship_notes || "",
  };

  if (!draft_answers.sponsor_status) missing_fields.push("Sponsor status");
  if (!draft_answers.relationship_type) missing_fields.push("Relationship type");
  if (!draft_answers.relationship_start_date) {
    missing_fields.push("Relationship start date");
  }
  if (
    draft_answers.relationship_type.toLowerCase().includes("spouse") &&
    !draft_answers.marriage_date
  ) {
    missing_fields.push("Marriage date");
  }
  if (!draft_answers.proof_of_relationship_notes) {
    missing_fields.push("Relationship evidence notes");
  }
  if (draft_answers.police_certificates_ready === "No") {
    missing_fields.push("Police certificate planning");
  }
  if (draft_answers.medicals_ready === "No") {
    missing_fields.push("Medical exam planning");
  }

  const preparation_notes = [
    "Ensure relationship history is consistent across all forms and evidence.",
    "Prepare a clean timeline of communication, visits, cohabitation, and major milestones.",
    "Review prior marriages, sponsorships, or dependent child issues carefully.",
  ];

  if (intake.dependent_children) {
    recommended_forms.push({
      form_key: "dependent_children_support",
      form_name: "Dependent children support package",
      status: "Conditional",
      notes: "Additional child-related supporting information may be required.",
    });
  }

  return {
    matter_type: "spousal_sponsorship",
    package_title: "Spousal Sponsorship Forms Assistant",
    recommended_forms,
    draft_answers,
    missing_fields,
    preparation_notes,
    summary:
      missing_fields.length === 0
        ? "The sponsorship forms package looks reasonably prepared from current intake."
        : "Some key sponsorship form fields still need to be completed or reviewed.",
  };
}

export function buildFormsAssistant(matterType, intake = {}, language = "en") {
  let result;

  if (matterType === "study_permit") {
    result = buildStudyPermitFormsAssistant(intake);
    return translateFormsAssistantValue(result, language);
  }

  if (matterType === "work_permit") {
    result = buildWorkPermitFormsAssistant(intake);
    return translateFormsAssistantValue(result, language);
  }

  if (matterType === "spousal_sponsorship") {
    result = buildSpousalSponsorshipFormsAssistant(intake);
    return translateFormsAssistantValue(result, language);
  }

  result = {
    matter_type: matterType || "unknown",
    package_title: "Forms Assistant",
    recommended_forms: [],
    draft_answers: {},
    missing_fields: ["Forms assistant has not been built for this matter type yet."],
    preparation_notes: [],
    summary: "No forms assistant is available for this matter type yet.",
  };

  return translateFormsAssistantValue(result, language);
}
